lib/src/binding_mqtt/mqtt_extensions.dart

```diff
@@ -0,0 +1,157 @@
+    final client = MqttServerClient.withPort(
+      _brokerAddress,
+      _id,
+      _mqttPort,
+    );
+
+    if (scheme == mqttSecureUriScheme) {
+      client.secure = true;
+    }
+
+    client
+      ..setProtocolV311()
+      ..keepAlivePeriod = 20;
+    return client;
+  }
+
+  String get _brokerAddress => host;
+
+  int get _mqttPort {
+    if (port == 0) {
+      return _defaultPort;
+    }
+
+    return port;
+  }
+
+  int get _defaultPort {
+    switch (scheme) {
+      case mqttUriScheme:
+        return defaultMqttPort;
+      case mqttSecureUriScheme:
+        return defaultMqttSecurePort;
+    }
+
+    throw StateError('MQTT URI scheme $scheme is not supported.');
+  }
+}
+
+/// Additional methods for making MQTT [Form]s easier to work with.
+extension MqttFormExtension on Form {
+  /// Indicates if this [Form] requires basic authentication.
+  bool requiresBasicAuthencation(BasicCredentials? credentials) {
+    if (_hasBasicSecurityScheme) {
+      return true;
+    }
+
+    return _hasAutoSecurityScheme && credentials == null;
+  }
+
+  T? _obtainVocabularyTerm<T>(String term) {
+    final curieString = mqttPrefixMapping.expandCurie(Curie(reference: term));
+
+    final value = additionalFields[curieString] ??
+        additionalFields['$defaultMqttPrefix:$term'];
+
+    if (value is T) {
+      return value;
+    }
+
+    return null;
+  }
+
+  /// Gets the MQTT topic for publishing from this [Form].
+  ///
+  /// Throws an [Exception] if no topic could be retrieved.
+  String get topicName {
+    final topic = _obtainVocabularyTerm<String>('topic');
+
+    if (topic == null) {
+      throw MqttBindingException('MQTT topic was not defined on form.');
+    }
+
+    return topic;
+  }
+
+  /// Gets the MQTT topic for subscribing from this [Form].
+  ///
+  /// Throws an [Exception] if no topic could be retrieved.
+  String get topicFilter {
+    final topic = _obtainVocabularyTerm<String>('filter');
+
+    if (topic == null) {
+      throw MqttBindingException('MQTT topic was not defined on form.');
+    }
+
+    return topic;
+  }
+
+  /// Gets the MQTT `retain` value from this [Form] if present.
+  ///
+  /// Returns `null` otherwise.
+  bool? get retain => _obtainVocabularyTerm<bool>('retain');
+
+  bool get _hasBasicSecurityScheme =>
+      securityDefinitions.whereType<BasicSecurityScheme>().isNotEmpty;
+
+  bool get _hasAutoSecurityScheme =>
+      securityDefinitions.whereType<AutoSecurityScheme>().isNotEmpty;
+
+  /// Gets the MQTT Quality of Service from this [Form] if present.
+  ///
+  /// Returns `null` otherwise.
+  MqttQos? get qualityOfService {
+    final qosValue = _obtainVocabularyTerm<String>('qos');
+    switch (qosValue) {
+      case 'quality:0':
+        return MqttQos.atMostOnce;
+      case 'quality:1':
+        return MqttQos.atLeastOnce;
+      case 'quality:2':
+        return MqttQos.exactlyOnce;
+    }
+
+    // TODO: Should a validation exception be thrown here instead?
```

Comment:
If a value other than `quality:{0,1,2}` is invalid as specified by the protocol, then yes, I think a validation exception should be thrown here. (If the `qosValue` is simply `null`, i.e., missing, returning `null` should still be fine, though.) Note that I don't have any experience with MQTT, this is just my intuitive judgement.

Reply:
Added a validation check in 09101eac77c25e0941130a9b42102e36730fd533 :) Only the three values should be valid (also see [here](https://w3c.github.io/wot-binding-templates/bindings/protocols/mqtt/index.html#qualityofservice)), so throwing an exception does make sense here, I think. I kept a TODO, though, as the validation could also be moved to the actual processing of the TD, but I'd tend to add this in a separate refactoring PR.